File: rag_backend/app/schemas/invite_code.py
"""
邀请码相关的Pydantic模型
"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, ConfigDict
from uuid import UUID


# =======================
# 邀请码创建相关
# =======================

class InviteCodeCreate(BaseModel):
    """创建邀请码的请求模型"""
    max_uses: int = Field(default=1, ge=1, le=100, description="最大使用次数")
    expires_hours: int = Field(default=24, ge=1, le=8760, description="过期时间(小时)")
    description: Optional[str] = Field(None, max_length=200, description="邀请描述")
    role: str = Field(default="member", description="被邀请用户角色")


class InviteCodeUpdate(BaseModel):
    """更新邀请码的请求模型"""
    is_active: Optional[bool] = Field(None, description="是否激活")
    description: Optional[str] = Field(None, max_length=200, description="邀请描述")


# =======================
# 邀请码响应相关
# =======================

class InviteCodeOut(BaseModel):
    """邀请码输出模型"""
    id: UUID
    code: str
    tenant_id: str
    created_by: UUID
    
    # 配置信息
    max_uses: int
    used_count: int
    expires_at: Optional[datetime]
    description: Optional[str]
    role: str
    
    # 状态信息
    is_active: bool
    is_expired: bool
    is_exhausted: bool
    is_valid: bool
    remaining_uses: int
    
    # 时间信息
    created_at: datetime
    updated_at: Optional[datetime]
    
    class Config:
        from_attributes = True


# 为了兼容性和统一命名，提供别名
InviteCodeResponse = InviteCodeOut


class InviteCodeSummary(BaseModel):
    """邀请码摘要信息"""
    id: UUID
    code: str
    description: Optional[str]
    max_uses: int
    used_count: int
    remaining_uses: int
    is_valid: bool
    expires_at: Optional[datetime]
    created_at: datetime
    
    model_config = ConfigDict(
        from_attributes=True
    )


# =======================
# 邀请码验证相关
# =======================

class InviteCodeValidation(BaseModel):
    """邀请码验证请求"""
    code: str = Field(..., min_length=8, max_length=32, description="邀请码")


class InviteCodeValidationResult(BaseModel):
    """邀请码验证结果"""
    valid: bool
    message: str
    tenant_id: Optional[str] = None
    company_name: Optional[str] = None
    creator_name: Optional[str] = None
    description: Optional[str] = None
    expires_at: Optional[datetime] = None
    remaining_uses: Optional[int] = None


# =======================
# 邀请码使用记录
# =======================

class InviteCodeUsageOut(BaseModel):
    """邀请码使用记录输出"""
    id: UUID
    invite_code_id: UUID
    user_id: UUID
    used_at: datetime
    ip_address: Optional[str]
    user_agent: Optional[str]
    
    # 关联信息
    user_email: Optional[str] = None
    user_name: Optional[str] = None
    
    model_config = ConfigDict(
        from_attributes=True
    )


# =======================
# 邀请码统计
# =======================

class InviteCodeStats(BaseModel):
    """邀请码统计信息"""
    total_codes: int
    active_codes: int
    expired_codes: int
    exhausted_codes: int
    total_uses: int
    total_invited_users: int


# =======================
# 批量操作
# =======================

class InviteCodeBatchCreate(BaseModel):
    """批量创建邀请码"""
    count: int = Field(..., ge=1, le=50, description="创建数量")
    max_uses: int = Field(default=1, ge=1, le=100, description="每个邀请码的最大使用次数")
    expires_hours: int = Field(default=24, ge=1, le=8760, description="过期时间(小时)")
    description_template: Optional[str] = Field(None, max_length=200, description="描述模板")
    role: str = Field(default="member", description="被邀请用户角色")


class InviteCodeBatchResult(BaseModel):
    """批量创建结果"""
    success: bool
    created_count: int
    codes: list[str]
    message: str

File: rag_frontend/src/api/invite-code.ts
import { request } from '@/utils/request'

export interface InviteCode {
  id: string
  code: string
  tenant_id: string
  created_by: string
  max_uses: number
  used_count: number
  expires_at: string
  description?: string
  role: string
  is_active: boolean
  is_expired: boolean
  is_exhausted: boolean
  is_valid: boolean
  remaining_uses: number
  created_at: string
  updated_at?: string
}

export interface InviteCodeStats {
  total_codes: number
  active_codes: number
  expired_codes: number
  exhausted_codes: number
  total_uses: number
  total_invited_users: number
}

export interface InviteCodeValidationResult {
  valid: boolean
  message: string
  tenant_id?: string
  company_name?: string
  creator_name?: string
  description?: string
  expires_at?: string
  remaining_uses?: number
}

export interface CreateInviteCodeRequest {
  max_uses: number
  expires_hours: number
  description?: string
  role?: string
}

export interface BatchCreateInviteCodeRequest {
  count: number
  max_uses: number
  expires_hours: number
  description_template?: string
  role?: string
}

export interface UpdateInviteCodeRequest {
  is_active?: boolean
  description?: string
}

export const inviteCodeApi = {
  async getCodes(params: {
    skip?: number
    limit?: number
    include_inactive?: boolean
  } = {}): Promise<InviteCode[]> {
    const { skip = 0, limit = 20, include_inactive = false } = params
    return request<InviteCode[]>(
      `/api/v1/invite-code?skip=${skip}&limit=${limit}&include_inactive=${include_inactive}`
    )
  },

  async getCode(codeId: string): Promise<InviteCode> {
    return request<InviteCode>(`/api/v1/invite-code/${codeId}`)
  },

  async createCode(data: CreateInviteCodeRequest): Promise<InviteCode> {
    return request<InviteCode>('/api/v1/invite-code', {
      method: 'POST',
      body: data,
    })
  },

  async batchCreateCodes(
    data: BatchCreateInviteCodeRequest
  ): Promise<InviteCode[]> {
    return request<InviteCode[]>('/api/v1/invite-code/batch', {
      method: 'POST',
      body: data,
    })
  },

  async updateCode(
    codeId: string,
    data: UpdateInviteCodeRequest
  ): Promise<InviteCode> {
    return request<InviteCode>(`/api/v1/invite-code/${codeId}`, {
      method: 'PUT',
      body: data,
    })
  },

  async deleteCode(code: string): Promise<void> {
    return request(`/api/v1/invite-code/${code}`, {
      method: 'DELETE',
    })
  },

  async getStats(): Promise<InviteCodeStats> {
    return request<InviteCodeStats>('/api/v1/invite-code/stats')
  },

  async validateCode(code: string): Promise<InviteCodeValidationResult> {
    return request<InviteCodeValidationResult>(
      `/api/v1/invite-code/validate/${code}`
    )
  },
}
